File: projects/gemini-cli/src/oauth/token.js
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * OAuth Token Handler
 * Handles OAuth token requests and responses
 */

/**
 * Parse OAuth token request
 * @param {string} requestStr - Token request string
 * @returns {object} Parsed token request
 */
export function parseOAuthTokenRequest(requestStr) {
  if (typeof requestStr !== 'string') {
    throw new TypeError('Request must be a string');
  }

  if (!requestStr.trim()) {
    throw new Error('Request cannot be empty');
  }

  try {
    // Try JSON first
    const request = JSON.parse(requestStr);

    // Validate OAuth 2.0 format
    if (!request.grant_type) {
      throw new Error('Missing grant_type parameter');
    }

    return request;
  } catch (error) {
    if (error instanceof SyntaxError) {
      throw new Error(`Invalid JSON: ${error.message}`);
    }
    throw error;
  }
}

/**
 * Validate OAuth token request
 * @param {object} request - Parsed token request
 * @returns {object} Validation result
 */
export function validateOAuthTokenRequest(request) {
  const errors = [];

  if (!request || typeof request !== 'object') {
    errors.push('Request must be an object');
    return { valid: false, errors };
  }

  // Validate grant type
  if (!request.grant_type) {
    errors.push('grant_type is required');
  } else {
    const validGrantTypes = [
      'authorization_code',
      'refresh_token',
      'client_credentials',
      'password'
    ];

    if (!validGrantTypes.includes(request.grant_type)) {
      errors.push(`Invalid grant_type: ${request.grant_type}`);
    }
  }

  // Validate based on grant type
  switch (request.grant_type) {
    case 'authorization_code':
      if (!request.code) {
        errors.push('authorization_code grant requires code parameter');
      }
      if (!request.redirect_uri) {
        errors.push('authorization_code grant requires redirect_uri parameter');
      }
      break;

    case 'refresh_token':
      if (!request.refresh_token) {
        errors.push('refresh_token grant requires refresh_token parameter');
      }
      break;

    case 'client_credentials':
      // Client credentials may not need additional parameters
      break;

    case 'password':
      if (!request.username) {
        errors.push('password grant requires username parameter');
      }
      if (!request.password) {
        errors.push('password grant requires password parameter');
      }
      break;
  }

  // Validate client credentials
  if (request.client_id && typeof request.client_id !== 'string') {
    errors.push('client_id must be a string');
  }

  if (request.client_secret && typeof request.client_secret !== 'string') {
    errors.push('client_secret must be a string');
  }

  // Check for malicious patterns in all string fields
  function checkMaliciousPatterns(obj, path = '') {
    if (typeof obj === 'string') {
      const maliciousPatterns = [
        '<script',
        'javascript:',
        'eval(',
        '../',
        '..\\',
        'UNION',
        'SELECT',
        'DROP',
        'DELETE',
        'INSERT',
        'UPDATE'
      ];

      for (const pattern of maliciousPatterns) {
        if (obj.includes(pattern)) {
          errors.push(`Malicious pattern "${pattern}" found in ${path || 'field'}`);
        }
      }
    } else if (typeof obj === 'object' && obj !== null) {
      for (const [key, value] of Object.entries(obj)) {
        const newPath = path ? `${path}.${key}` : key;
        checkMaliciousPatterns(value, newPath);
      }
    }
  }

  checkMaliciousPatterns(request);

  return {
    valid: errors.length === 0,
    errors
  };
}

/**
 * Parse OAuth token response
 * @param {string} responseStr - Token response string
 * @returns {object} Parsed token response
 */
export function parseOAuthTokenResponse(responseStr) {
  if (typeof responseStr !== 'string') {
    throw new TypeError('Response must be a string');
  }

  if (!responseStr.trim()) {
    throw new Error('Response cannot be empty');
  }

  try {
    const response = JSON.parse(responseStr);

    // Validate OAuth 2.0 response format
    if (response.error) {
      // Error response
      if (!response.error_description && !response.error_uri) {
        // At least one of these should be present
      }
    } else {
      // Success response should have access_token
      if (!response.access_token) {
        throw new Error('Success response must contain access_token');
      }
    }

    return response;
  } catch (error) {
    if (error instanceof SyntaxError) {
      throw new Error(`Invalid JSON: ${error.message}`);
    }
    throw error;
  }
}

/**
 * Validate OAuth token response
 * @param {object} response - Parsed token response
 * @returns {object} Validation result
 */
export function validateOAuthTokenResponse(response) {
  const errors = [];

  if (!response || typeof response !== 'object') {
    errors.push('Response must be an object');
    return { valid: false, errors };
  }

  if (response.error) {
    // Error response validation
    if (typeof response.error !== 'string') {
      errors.push('error must be a string');
    }

    if (response.error_description && typeof response.error_description !== 'string') {
      errors.push('error_description must be a string');
    }

    if (response.error_uri && typeof response.error_uri !== 'string') {
      errors.push('error_uri must be a string');
    }
  } else {
    // Success response validation
    if (!response.access_token || typeof response.access_token !== 'string') {
      errors.push('access_token is required and must be a string');
    }

    if (!response.token_type || typeof response.token_type !== 'string') {
      errors.push('token_type is required and must be a string');
    }

    if (response.expires_in !== undefined) {
      if (typeof response.expires_in !== 'number' || response.expires_in <= 0) {
        errors.push('expires_in must be a positive number');
      }
    }

    if (response.refresh_token && typeof response.refresh_token !== 'string') {
      errors.push('refresh_token must be a string');
    }

    if (response.scope && typeof response.scope !== 'string') {
      errors.push('scope must be a string');
    }
  }

  return {
    valid: errors.length === 0,
    errors
  };
}

/**
 * Create OAuth token request
 * @param {object} params - Request parameters
 * @returns {object} OAuth token request
 */
export function createOAuthTokenRequest(params) {
  if (!params || typeof params !== 'object') {
    throw new Error('Parameters are required');
  }

  if (!params.grant_type) {
    throw new Error('grant_type is required');
  }

  const request = {
    grant_type: params.grant_type
  };

  // Add optional parameters
  if (params.code) request.code = params.code;
  if (params.redirect_uri) request.redirect_uri = params.redirect_uri;
  if (params.refresh_token) request.refresh_token = params.refresh_token;
  if (params.client_id) request.client_id = params.client_id;
  if (params.client_secret) request.client_secret = params.client_secret;
  if (params.username) request.username = params.username;
  if (params.password) request.password = params.password;
  if (params.scope) request.scope = params.scope;

  return request;
}

/**
 * Create OAuth token response
 * @param {string} accessToken - Access token
 * @param {string} tokenType - Token type
 * @param {object} options - Additional options
 * @returns {object} OAuth token response
 */
export function createOAuthTokenResponse(accessToken, tokenType, options = {}) {
  if (!accessToken || typeof accessToken !== 'string') {
    throw new Error('access_token is required and must be a string');
  }

  if (!tokenType || typeof tokenType !== 'string') {
    throw new Error('token_type is required and must be a string');
  }

  const response = {
    access_token: accessToken,
    token_type: tokenType
  };

  // Add optional parameters
  if (options.expires_in !== undefined) {
    if (typeof options.expires_in === 'number' && options.expires_in > 0) {
      response.expires_in = options.expires_in;
    }
  }

  if (options.refresh_token) {
    response.refresh_token = options.refresh_token;
  }

  if (options.scope) {
    response.scope = options.scope;
  }

  return response;
}

File: projects/gemini-cli/gofuzz/internal/oauth/token.go
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package oauth

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"strings"
)

// OAuth token structures for fuzzing token handling logic
type TokenResponse struct {
	AccessToken  string `json:"access_token"`
	TokenType    string `json:"token_type"`
	ExpiresIn    int    `json:"expires_in"`
	RefreshToken string `json:"refresh_token,omitempty"`
	Scope        string `json:"scope,omitempty"`
	IDToken      string `json:"id_token,omitempty"`
}

type TokenRequest struct {
	GrantType    string `json:"grant_type"`
	Code         string `json:"code,omitempty"`
	RedirectURI  string `json:"redirect_uri,omitempty"`
	ClientID     string `json:"client_id,omitempty"`
	ClientSecret string `json:"client_secret,omitempty"`
	RefreshToken string `json:"refresh_token,omitempty"`
	Scope        string `json:"scope,omitempty"`
}

// ValidateTokenResponse checks OAuth token response for security issues
func ValidateTokenResponse(data []byte) (*TokenResponse, error) {
	if len(data) > 64*1024 { // 64KB limit
		return nil, errors.New("token response too large")
	}

	var token TokenResponse
	if err := json.NewDecoder(bytesLimited(data, 1<<20)).Decode(&token); err != nil {
		return nil, fmt.Errorf("invalid token JSON: %w", err)
	}

	// Validate access token
	if err := validateToken(token.AccessToken, "access_token"); err != nil {
		return nil, err
	}

	// Validate token type
	if token.TokenType != "" {
		validTypes := []string{"Bearer", "bearer", "MAC", "mac"}
		valid := false
		for _, t := range validTypes {
			if token.TokenType == t {
				valid = true
				break
			}
		}
		if !valid {
			return nil, fmt.Errorf("invalid token_type: %s", token.TokenType)
		}
	}

	// Validate expires_in
	if token.ExpiresIn < 0 || token.ExpiresIn > 86400*365 { // Max 1 year
		return nil, errors.New("invalid expires_in value")
	}

	// Validate refresh token if present
	if token.RefreshToken != "" {
		if err := validateToken(token.RefreshToken, "refresh_token"); err != nil {
			return nil, err
		}
	}

	// Validate ID token if present (basic JWT structure check)
	if token.IDToken != "" {
		if err := validateJWTStructure(token.IDToken); err != nil {
			return nil, fmt.Errorf("invalid id_token: %w", err)
		}
	}

	return &token, nil
}

// ValidateTokenRequest checks OAuth token request for security issues
func ValidateTokenRequest(data []byte) (*TokenRequest, error) {
	if len(data) > 32*1024 { // 32KB limit
		return nil, errors.New("token request too large")
	}

	var req TokenRequest
	if err := json.NewDecoder(bytesLimited(data, 1<<20)).Decode(&req); err != nil {
		return nil, fmt.Errorf("invalid request JSON: %w", err)
	}

	// Validate grant type
	validGrants := []string{
		"authorization_code",
		"refresh_token",
		"client_credentials",
		"password", // Though not recommended
	}
	valid := false
	for _, g := range validGrants {
		if req.GrantType == g {
			valid = true
			break
		}
	}
	if !valid {
		return nil, fmt.Errorf("invalid grant_type: %s", req.GrantType)
	}

	// Validate redirect URI format
	if req.RedirectURI != "" {
		if err := validateRedirectURI(req.RedirectURI); err != nil {
			return nil, err
		}
	}

	// Check for potential injection in client credentials
	if strings.ContainsAny(req.ClientID, "<>\"'&;") {
		return nil, errors.New("potentially dangerous characters in client_id")
	}

	return &req, nil
}

func validateToken(token, tokenType string) error {
	if len(token) == 0 {
		return fmt.Errorf("%s cannot be empty", tokenType)
	}

	if len(token) > 8192 { // 8KB max token size
		return fmt.Errorf("%s too long", tokenType)
	}

	// Check for obvious injection attempts
	dangerous := []string{"<script", "javascript:", "data:", "vbscript:", "onload="}
	tokenLower := strings.ToLower(token)
	for _, pattern := range dangerous {
		if strings.Contains(tokenLower, pattern) {
			return fmt.Errorf("potentially dangerous pattern in %s", tokenType)
		}
	}

	return nil
}

func validateJWTStructure(token string) error {
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return errors.New("JWT must have 3 parts separated by dots")
	}

	// Basic length checks for each part
	for i, part := range parts {
		if len(part) == 0 {
			return fmt.Errorf("JWT part %d cannot be empty", i+1)
		}
		if len(part) > 16384 { // 16KB per part max
			return fmt.Errorf("JWT part %d too long", i+1)
		}
	}

	return nil
}

func validateRedirectURI(uri string) error {
	if len(uri) > 2048 {
		return errors.New("redirect_uri too long")
	}

	// Must be HTTPS or localhost for security
	if !strings.HasPrefix(uri, "https://") &&
		!strings.HasPrefix(uri, "http://localhost") &&
		!strings.HasPrefix(uri, "http://127.0.0.1") {
		return errors.New("redirect_uri must use HTTPS or localhost")
	}

	// Check for dangerous characters
	if strings.ContainsAny(uri, "<>\"'") {
		return errors.New("redirect_uri contains dangerous characters")
	}

	return nil
}

// bytesLimited returns an io.Reader capped to at most cap bytes and ends with io.EOF.
func bytesLimited(b []byte, cap int64) io.Reader {
	if int64(len(b)) > cap {
		b = b[:cap]
	}
	return bytes.NewReader(b)
}
